### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by pepeiborra**
* Fixed build with GHC 8, but the instrumentation in the Hoed case does not type check. This is presumably due to flaky support for generation of type signatures inside TH declaration splices.

* Refactored the Variables backend to a module and renamed Debug.Record to Debug.DebugTrace (couldn't think of any better name).

* Other minor refactorings and fixes.

### CODE REVIEW

File: src/Debug/Hoed.hs [hunk 1]
```diff
@@ -93,6 +95,8 @@ import           Language.Haskell.TH
 import           Language.Haskell.TH.Syntax
 import           System.Clock
 
+{-# ANN module ("hlint: ignore Redundant bracket" :: String) #-}
```
Comment (ndmitchell) [inline src/Debug/Hoed.hs]: What triggers this warning? Could you not just remove the redundant bracket instead?

**Comment by ndmitchell** [comment]
Looks good, although I'll wait for the CI's to confirm.

File: src/Debug/Hoed.hs (hunk 1 above)
Comment (pepeiborra) [reply src/Debug/Hoed.hs]: Conditional compilation:
```
      ([ ViewPatterns
       , PartialTypeSignatures
       , ExtendedDefaultRules
       , FlexibleContexts
       ]
#if __GLASGOW_HASKELL__ >= 802
       ++
       [DeriveAnyClass | generateObservableInstances] ++
       [DerivingStrategies | generateObservableInstances] ++
       [DeriveGeneric | generateGenericInstances]
#endif
      )
``

Comment (ndmitchell) [reply src/Debug/Hoed.hs]: Yuk, but fair enough.

**Comment by pepeiborra** [comment]
These are all due to conditional compilation.

**Comment by ndmitchell** [comment]
If you add them as ignored in the weeder config file, then it should pass.

**Comment by pepeiborra** [comment]
I've restored the trace regression test in 8.0, should be fine now but let's wait and see

**Pull request merged**